clippy_lints/src/utils/mod.rs

```diff
@@ -97,19 +98,90 @@ pub fn in_macro(span: Span) -> bool {
 /// Used to store the absolute path to a type.
 ///
 /// See `match_def_path` for usage.
-#[derive(Debug)]
-pub struct AbsolutePathBuffer {
-    pub names: Vec<symbol::LocalInternedString>,
+pub struct AbsolutePathPrinter<'a, 'tcx> {
+    pub tcx: TyCtxt<'a, 'tcx, 'tcx>,
 }
 
-impl ty::item_path::ItemPathBuffer for AbsolutePathBuffer {
-    fn root_mode(&self) -> &ty::item_path::RootMode {
-        const ABSOLUTE: &ty::item_path::RootMode = &ty::item_path::RootMode::Absolute;
-        ABSOLUTE
+use rustc::ty::print::Printer;
+
+#[allow(clippy::diverging_sub_expression)]
```

Comment:
This sounds like a false positive - is an issue filled?

Reply:
Not really a FP:
Since `type Error = !` and we call `let mut path = print_prefix(self)?;`, the `?` is a diverging sub-expression. IMO allowing this here is ok.

Reply:
I disagree, `?` with a diverging error path should not be warned against, the behavior is more like that of `let Ok(mut path) = print_prefix(self);`.

cc @Manishearth @oli-obk